Add datastore latency logic

lyddonb opened the pull request:
The datastore hook will now also leverage the latency configuration if
its set to insert latency spikes based off that configuration.

Also updated the README documentation to include the latency docs. As
well as some other quick tweaks to the README.

The datastore configuration also needed updating:
- Add LatencyConfig and LatenciesConfig objects in hydrate for storing
latency configurations in memory.
- Add a default latency configuration to part of the datastore configration.
- Also added flags for latency and errors so you can do them optionally

On the way to supporting latency we also refactored the hooks
- Move some of the more reusable error hook code into it's own hook
module.
- This includes moving the chance/roll logic out to it's own module under
gchaos and exposing a roll function in that module.

@RealKinetic/devs

lyddonb committed:
Add support for latency configuration

Add LatencyConfig and LatenciesConfig objects in hydrate for storing
latency configurations in memory.

Add a default latency configuration to part of the datastore
configration.

lyddonb committed:
Refactor hooks

Move some of the more reusable error hook code into it's own hook
module. This way we can reuse it for the latency code that's coming.

This includes moving the chance/roll logic out to it's own module under
gchaos and exposing a roll function in that module.

lyddonb committed:
Add datastore latency logic

The datastore hook will now also leverage the latency configuration if
its set to insert latency spikes based off that configuration.

Also added flags for latency and errors so you can do them optionally

Also updated the README documentation to include the latency docs. As
well as some other quick tweaks.

lyddonb committed:
Update latency.py

Kill unnecessary TODO

robertkluin commented on README.md:
```diff
@@ -20,39 +20,93 @@ host.
+            "GET": ({ 'google.appengine.api.datastore_errors.Timeout': 1 }, 0.01),
+            "PUT": (
+                { 'google.appengine.api.datastore_errors.BadRequestError': 0.75
+                'google.appengine.api.datastore_errors.InternalError': 0.25
```
This missing a comma?

robertkluin commented on README.md:
```diff
@@ -20,39 +20,93 @@ host.
 
-The list of supported Google Datastore errors can be found here:
+You can disable the entire datastore configuration. If you set the `enabled` flag
+to `False` nothing will happer. If it's set to true it will then look for errors
```
happen

robertkluin commented on README.md:
```diff
@@ -20,39 +20,93 @@ host.
+which means nothing will happen. Where the maximum is `1.00` which means an error
+or latency spike is guranteed to happen. Otherwise it is just probabistic. Also
+each action is independent of the previous action. Meaning that we don't increase
+or decrease the likelyhood of an action being trigger based off subsequent actions.
```
likelihood   triggered

robertkluin commented on README.md:
```diff
@@ -20,39 +20,93 @@ host.
 
+For each entry in the dictionary you then add a corresponding integer for the
+error to be hit. The entries should add up to 100. This way you can say a specific
+error should have a 50% likely hood of being hit by setting it's value to `50`
```
probability

lyddonb committed:
Tick the todo column :)

robertkluin commented on gchaos/config/hydrate.py:
```diff
@@ -63,31 +65,35 @@ def __init__(self, config):
             *config.get("DELETE", DEFAULT_ERROR_ENTRY))
         self.get_errors = ErrorConfig(*config.get("GET", DEFAULT_ERROR_ENTRY))
         self.put_errors = ErrorConfig(*config.get("PUT", DEFAULT_ERROR_ENTRY))
+        # self.enabled =
```
delete this line?

robertkluin commented on gchaos/gae/datastore/latency.py:
```diff
@@ -0,0 +1,132 @@
+
+def stall(latency):
+    """Based off the latency stall for that long. The latency is a tuple if only
+    one value is provided then stall for exactly that long. If to values are
```
two

robertkluin reviewed (approved):
Seems reasonable to me.

njoyce commented on gchaos/chance.py:
```diff
@@ -0,0 +1,52 @@
+    Return:
+        int
+    """
+    random.seed()
```
Shouldn't need to seed every call

njoyce commented on gchaos/gae/datastore/latency.py:
```diff
@@ -0,0 +1,132 @@
+    if max_ < min_:
+        raise InvalidLatencyException(latency)
+
+    random.seed()
```
Do we need to seed every call?

lyddonb committed:
Move random.seed to top level

Move the random.seed calls from chance and latency to the init method.

Also cleaning up some code review spelling mistakes.

lyddonb commented:
@njoyce @robertkluin Comments addressed
